stan/math/rev/core/vari.hpp

```diff
@@ -197,47 +217,49 @@ using vari = vari_value<double>;
  * derivative with respect to the root of the derivative tree.
  *
  */
-template <typename T>
-class vari_value<T, std::enable_if_t<is_eigen_dense_base<T>::value>>
-    : public vari_base {
+template <typename T_val, typename T_adj>
+class vari_value<T_val, T_adj,
+                 std::enable_if_t<is_eigen_dense_base<T_val>::value
+                                  && is_eigen_dense_base<T_adj>::value>>
+    : public get_vari_base<
+          std::is_trivially_destructible<T_val>::value
+          && std::is_trivially_destructible<T_adj>::value>::type {
+  using Base = typename get_vari_base<
+      std::is_trivially_destructible<T_val>::value
+      && std::is_trivially_destructible<T_adj>::value>::type;
+
  public:
   static_assert(
-      is_plain_type<T>::value,
-      "The template for this var is an"
-      " expression but a var_value's inner type must be assignable such as"
-      " a double, Eigen::Matrix, or Eigen::Array");
+      std::is_same<std::decay_t<plain_type_t<T_val>>,
+                   std::decay_t<plain_type_t<T_adj>>>::value,
+      "The templates for value and adjoint must have same plain type!");
+  static_assert(
+      std::is_same<std::decay_t<T_val>, std::decay_t<ref_type_t<T_val>>>::value
+          && std::is_same<std::decay_t<T_adj>,
+                          std::decay_t<ref_type_t<T_adj>>>::value,
```

Comment:
We don't want to allow just any expression to be used in vari (it would make no sense to have adjoint of type sum of two matrices). Eventually I'd like to use general expressions in vari, but that will require quite some more work. 

So this checks that (decayed) `T` and `ref_type__t<T>` are same, which is true only for expressions that allow direct access to their data, such as block etc.